include/nbl/system/IAsyncQueueDispatcher.h

```diff
@@ -0,0 +1,58 @@
+#ifndef __NBL_I_ASYNC_QUEUE_DISPATCHER_H_INCLUDED__
+#define __NBL_I_ASYNC_QUEUE_DISPATCHER_H_INCLUDED__
+
+#include "nbl/system/IThreadHandler.h"
+#include "nbl/core/Types.h"
+
+namespace nbl {
+namespace system
+{
+
+template <typename QueueElementType, typename InternalStateType>
+class IAsyncQueueDispatcher : public IThreadHandler<InternalStateType>
+{
+    using base_t = system::IThreadHandler<InternalStateType>;
+
+public:
+    void enqueue(uint32_t count, const queue_element_t* elements)
+    {
+        auto raii_handler = base_t::createRAIIDispatchHandler();
+
+        for (uint32_t i = 0u; i < count; ++i)
+        {
+            auto e = elements[i];
+            q.push(std::move(e));
+        }
```

Comment:
a cool trick when using a fixed size circular buffer ( https://github.com/Devsh-Graphics-Programming/Nabla/pull/28/files#r569615420 ) would be to check if you can push (current size is `<(capacity-1)` because `work` could be using a reference to the old tail of the queue)

If you can you push, if you can't you wait on a **second condition variable** which gets notified when the `work_epilogue` signals it

Reply:
then you never overflow, but just stall when buffer is full